sklearn/ensemble/_gb.py

```diff
@@ -1211,20 +1477,60 @@ def __init__(
             ccp_alpha=ccp_alpha,
         )
 
-    def _validate_y(self, y, sample_weight):
+    def _encode_y(self, y, sample_weight):
+        # encode classes into 0 ... n_classes - 1 and sets attributes classes_
+        # and n_trees_per_iteration_
         check_classification_targets(y)
-        self.classes_, y = np.unique(y, return_inverse=True)
-        n_trim_classes = np.count_nonzero(np.bincount(y, sample_weight))
+
+        label_encoder = LabelEncoder()
+        encoded_y = label_encoder.fit_transform(y)
+        self.classes_ = label_encoder.classes_
+        n_classes = self.classes_.shape[0]
+        # only 1 tree for binary classification. For multiclass classification,
+        # we build 1 tree per class.
+        self.n_trees_per_iteration_ = 1 if n_classes <= 2 else n_classes
+        encoded_y = encoded_y.astype(float, copy=False)
+
+        # From here on, it is additional to the HGBT case.
+        # expose n_classes_ attribute
+        self.n_classes_ = n_classes
+        if sample_weight is None:
+            n_trim_classes = n_classes
+        else:
+            n_trim_classes = np.count_nonzero(
+                np.bincount(encoded_y.astype(int), sample_weight)
```

Comment:
A few lines above where the `LabelEncoder` is called, can the variable be called `encoded_y_int`:

```python
label_encoder = LabelEncoder()
encoded_y_int = label_encoder.fit_transform(y)
```

This way, we can reuse it here without having to convert `encoded_y` back into an int.

Reply:
good point